zcash_client_backend/src/encoding.rs

```diff
@@ -0,0 +1,173 @@
+    let mut data: Vec<u8> = vec![];
+    write(&mut data).expect("Should be able to write to a Vec");
+
+    let converted =
+        convert_bits(&data, 8, 5, true).expect("Should be able to convert Vec<u8> to Vec<u5>");
+    let encoded = Bech32::new_check_data(hrp.into(), converted).expect("hrp is not empty");
+
+    encoded.to_string()
+}
+
+fn bech32_decode<T, F>(hrp: &str, s: &str, read: F) -> Result<Option<T>, Error>
+where
+    F: Fn(Vec<u8>) -> Option<T>,
+{
+    let decoded = s.parse::<Bech32>()?;
+    if decoded.hrp() == hrp {
+        convert_bits(decoded.data(), 5, 8, false).map(|data| read(data))
+    } else {
+        Ok(None)
+    }
+}
+
+/// Writes a [`PaymentAddress`] as a Bech32-encoded string.
+///
+/// # Examples
+///
+/// ```
+/// use pairing::bls12_381::Bls12;
+/// use rand::{SeedableRng, XorShiftRng};
+/// use sapling_crypto::{
+///     jubjub::edwards,
+///     primitives::{Diversifier, PaymentAddress},
+/// };
+/// use zcash_client_backend::{
+///     constants::testnet::HRP_SAPLING_PAYMENT_ADDRESS,
+///     encoding::encode_payment_address,
+/// };
+/// use zcash_primitives::JUBJUB;
+///
+/// let rng = &mut XorShiftRng::from_seed([0x3dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
+///
+/// let pa = PaymentAddress {
+///     diversifier: Diversifier([0u8; 11]),
+///     pk_d: edwards::Point::<Bls12, _>::rand(rng, &JUBJUB).mul_by_cofactor(&JUBJUB),
+/// };
+///
+/// assert_eq!(
+///     encode_payment_address(HRP_SAPLING_PAYMENT_ADDRESS, &pa),
+///     "ztestsapling1qqqqqqqqqqqqqqqqqqxrrfaccydp867g6zg7ne5ht37z38jtfyw0ygmp0ja6hhf07twjq6awtaj",
+/// );
+/// ```
+pub fn encode_payment_address(hrp: &str, addr: &PaymentAddress<Bls12>) -> String {
+    bech32_encode(hrp, |w| {
+        w.write_all(&addr.diversifier.0)?;
+        addr.pk_d.write(w)
+    })
+}
+
+/// Decodes a [`PaymentAddress`] from a Bech32-encoded string.
+///
+/// # Examples
+///
+/// ```
+/// use pairing::bls12_381::Bls12;
+/// use rand::{SeedableRng, XorShiftRng};
+/// use sapling_crypto::{
+///     jubjub::edwards,
+///     primitives::{Diversifier, PaymentAddress},
+/// };
+/// use zcash_client_backend::{
+///     constants::testnet::HRP_SAPLING_PAYMENT_ADDRESS,
+///     encoding::decode_payment_address,
+/// };
+/// use zcash_primitives::JUBJUB;
+///
+/// let rng = &mut XorShiftRng::from_seed([0x3dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
+///
+/// let pa = PaymentAddress {
+///     diversifier: Diversifier([0u8; 11]),
+///     pk_d: edwards::Point::<Bls12, _>::rand(rng, &JUBJUB).mul_by_cofactor(&JUBJUB),
+/// };
+///
+/// assert_eq!(
+///     decode_payment_address(
+///         HRP_SAPLING_PAYMENT_ADDRESS,
+///         "ztestsapling1qqqqqqqqqqqqqqqqqqxrrfaccydp867g6zg7ne5ht37z38jtfyw0ygmp0ja6hhf07twjq6awtaj",
+///     ),
+///     Ok(Some(pa)),
+/// );
+/// ```
+pub fn decode_payment_address(hrp: &str, s: &str) -> Result<Option<PaymentAddress<Bls12>>, Error> {
+    bech32_decode(hrp, s, |data| {
+        let mut diversifier = Diversifier([0; 11]);
```

Comment:
I'm not sure, but are we supposed to check here that the diversifier is valid before returning a `PaymentAddress` struct?

Reply:
Done.